File: test/javascript/plugins.js
// plugins.js -- exhaustive smoke tests covering every wasm-js side-loaded
// plugin.  Each test exercises a different plugin via its plugin-name
// string; the side modules ship preloaded into MEMFS by
// swig/wasm-js/CMakeLists.txt's wasmjs_side_plugin() invocations.
//
// The 21 plugins covered:
//   conic:       qrqp, ipqp, nlpsol (qp_to_nlp), highs
//   nlpsol:      sqpmethod, qrsqp, scpgen, feasiblesqpmethod, ipopt
//   integrator:  rk, collocation
//   linsol:      qr, ldl, tridiag, lsqr, symbolicqr
//   interpolant: linear, bspline
//   rootfinder:  newton, fast_newton, nlpsol (implicit_to_nlp)
//
// Most tests deliberately keep the problem trivial -- the goal is plugin
// coverage, not algorithm validation (per-algorithm tests live in the
// dedicated test files conic.js / integration.js / etc.).

const { assertEqual, assertTrue, assertAlmost, assertArrayAlmostEqual,
        runTests } = require("./_helpers");

// ---------------------------------------------------------------------
// linsol plugins:  Ax = b  with A = diag([1,2,3]) -> x = b ./ [1,2,3]
// ---------------------------------------------------------------------
function _linsol(M, plugin) {
  const A = M.DM([[1, 0, 0], [0, 2, 0], [0, 0, 3]]);
  const ls = M.Linsol("ls", plugin, A.sparsity());
  ls.sfact(A);
  ls.nfact(A);
  const b = M.DM([6, 8, 9]);
  const x = ls.solve(A, b, false);
  assertArrayAlmostEqual(x.nonzeros(), [6, 4, 3], 9, plugin + ": x = b/[1,2,3]");
}
function test_linsol_qr        (M) { _linsol(M, "qr"); }
function test_linsol_ldl       (M) {
  // ldl is symmetric-only; symmetric A = diag works.
  _linsol(M, "ldl");
}
function test_linsol_lsqr      (M) {
  // lsqr is iterative; same trivial diag input works.
  _linsol(M, "lsqr");
}
function test_linsol_symbolicqr(M) { _linsol(M, "symbolicqr"); }
function test_linsol_tridiag   (M) {
  // Tridiagonal solver: needs ACTUAL tri-banded (3 nonzeros per
  // interior column).  linsol_tridiag.cpp's solve() indexes
  // A[a_colind[i] + {0,1,2}] for sub/main/super-diag, so a diag-only
  // sparsity reads out of bounds and produces garbage.
  // Build a 3x3 symmetric tridiag pattern via triplet (col-major fill).
  // Pattern (rows, cols): (0,0),(1,0),(0,1),(1,1),(2,1),(1,2),(2,2).
  const rows = [0n, 1n, 0n, 1n, 2n, 1n, 2n];
  const cols = [0n, 0n, 1n, 1n, 1n, 2n, 2n];
  const sp = M.Sparsity.triplet(3n, 3n, rows, cols);
  // A = [[2,1,0],[1,2,1],[0,1,2]], values in CSC order.
  const A = M.DM(sp, [2, 1, 1, 2, 1, 1, 2]);
  const ls = M.Linsol("ls", "tridiag", sp);
  ls.sfact(A);
  ls.nfact(A);
  // Pick b so that x = [1, 2, 3]: A*x = [2+2, 1+4+3, 2+6] = [4, 8, 8].
  const b = M.DM([4, 8, 8]);
  const x = ls.solve(A, b, false);
  assertArrayAlmostEqual(x.nonzeros(), [1, 2, 3], 6, "tridiag x");
}

// ---------------------------------------------------------------------
// integrator plugins
// ---------------------------------------------------------------------
function _integrator(M, plugin) {
  // dx/dt = x, x(0)=1 -> x(1) = e.
  const x   = M.SX.sym("x");
  const dae = { x: x, ode: x };
  const F   = M.integrator("F", plugin, dae);
  const r   = F({ x0: M.DM(1) });
  assertAlmost(Number(r.xf.nonzeros()[0]), Math.E, 3,
    plugin + ": x(1) ~ e");
}
function test_integrator_rk         (M) { _integrator(M, "rk"); }
function test_integrator_collocation(M) {
  // Collocation defaults to implicit Radau IIA; same trivial ODE.
  // Higher-order, so accuracy is much better than rk's default 4-step.
  const x   = M.SX.sym("x");
  const dae = { x: x, ode: x };
  const F   = M.integrator("F", "collocation", dae);
  const r   = F({ x0: M.DM(1) });
  assertAlmost(Number(r.xf.nonzeros()[0]), Math.E, 4,
    "collocation: x(1) ~ e");
}

// ---------------------------------------------------------------------
// interpolant plugins
// ---------------------------------------------------------------------
function test_interpolant_linear(M) {
  // 1-D linear interpolant of f(0)=0, f(1)=1, f(2)=4, f(3)=9.
  const grid   = [[0, 1, 2, 3]];
  const values = [0, 1, 4, 9];
  const f = M.interpolant("f", "linear", grid, values);
  // Mid-point of [1, 2] -> linear blend = (1+4)/2 = 2.5.
  const r1 = f(M.DM(1.5));
  assertAlmost(Number(r1.nonzeros()[0]), 2.5, 9, "linear mid 1.5");
  // Knot point: exact.
  const r2 = f(M.DM(2));
  assertAlmost(Number(r2.nonzeros()[0]), 4, 9, "linear at knot 2");
}
function test_interpolant_bspline(M) {
  // 1-D B-spline of the same 4-point sample.  Default degree 3.
  // Don't check exact mid-point match (bspline is smooth, not piecewise
  // linear); just verify the call succeeds and recovers knot values
  // approximately.  Bspline knots use clamped boundary conditions, so
  // f(0) = 0, f(3) = 9 exactly.
  const grid   = [[0, 1, 2, 3]];
  const values = [0, 1, 4, 9];
  const f = M.interpolant("f", "bspline", grid, values);
  const r0 = f(M.DM(0));
  assertAlmost(Number(r0.nonzeros()[0]), 0, 6, "bspline at left knot");
  const r3 = f(M.DM(3));
  assertAlmost(Number(r3.nonzeros()[0]), 9, 6, "bspline at right knot");
}

// ---------------------------------------------------------------------
// rootfinder plugins:  find x s.t. f(x) = 0
// ---------------------------------------------------------------------
function _rootfinder(M, plugin) {
  // Solve x^2 - 2 = 0 starting from x0 = 1.5 -> root ~ 1.41421356.
  const x = M.SX.sym("x");
  const f = M.Function("f", [x], [M.minus(M.times(x, x), M.SX(2))]);
  const opts = {};
  // implicit_to_nlp's "nlpsol" plugin requires a sub-NLP solver;
  // route through ipopt.
  if (plugin === "nlpsol") opts["nlpsol"] = "ipopt";
  const F = M.rootfinder("F", plugin, f, opts);
  const r = F(M.DM(1.5));
  assertAlmost(Number(r.nonzeros()[0]), Math.SQRT2, 6,
    plugin + ": sqrt(2)");
}
function test_rootfinder_newton     (M) { _rootfinder(M, "newton"); }
function test_rootfinder_fast_newton(M) { _rootfinder(M, "fast_newton"); }
function test_rootfinder_implicit_to_nlp(M) {
  // implicit_to_nlp wraps an Nlpsol; test only if ipopt is present.
  if (!M.has_nlpsol("ipopt")) {
    console.log("    (skip -- ipopt not loaded)");
    return;
  }
  _rootfinder(M, "nlpsol");
}

// ---------------------------------------------------------------------
// conic plugins:  min 0.5 x'Hx + g'x  subject to lba <= Ax <= uba
// ---------------------------------------------------------------------
function _conic(M, plugin, opts) {
  // Same simple QP used in conic.js's test_general_unconstrained.
  // min 0.5 (x^2 + y^2) - 0.7 x - 2.3 y -> [0.7, 2.3].
  const H = M.DM([[1, 0], [0, 1]]);
  const G = M.DM([[-0.7], [-2.3]]);
  const inf = Infinity;
  const LBX = M.DM([[-inf], [-inf]]);
  const UBX = M.DM([[inf], [inf]]);
  const qp = { h: H.sparsity() };
  const solver = M.conic("solver", plugin, qp, opts || {});
  const res = solver.call({ h: H, g: G, lbx: LBX, ubx: UBX });
  assertArrayAlmostEqual(res.x.nonzeros(), [0.7, 2.3], 4,
    plugin + ": x");
}
function test_conic_qrqp (M) { _conic(M, "qrqp",  { print_iter: false, print_header: false }); }
function test_conic_ipqp (M) { _conic(M, "ipqp",  { print_iter: false, print_header: false }); }
function test_conic_nlpsol(M) {
  // qp_to_nlp wrapper requires an Nlpsol; route through ipopt.
  if (!M.has_nlpsol("ipopt")) {
    console.log("    (skip -- ipopt not loaded)");
    return;
  }
  _conic(M, "nlpsol", { nlpsol: "ipopt", nlpsol_options: { ipopt: { print_level: 0 }, print_time: false } });
}
function test_conic_highs(M) { _conic(M, "highs", { highs: { output_flag: false } }); }

// ---------------------------------------------------------------------
// nlpsol plugins:  min f(x) s.t. lbg <= g(x) <= ubg
// ---------------------------------------------------------------------
function _nlpsol(M, plugin, opts) {
  // Rosenbrock-like simple convex bowl: f(x,y) = (x-1)^2 + (y-2)^2.
  // Optimum at (1, 2) with f* = 0.  No constraints.
  const xy = M.SX.sym("xy", 2);
  const x  = xy.get(false, M.Slice(0n, false));
  const y  = xy.get(false, M.Slice(1n, false));
  const f  = M.plus(M.times(M.minus(x, M.SX(1)), M.minus(x, M.SX(1))),
                    M.times(M.minus(y, M.SX(2)), M.minus(y, M.SX(2))));
  const nlp = { x: xy, f: f };
  const solver = M.nlpsol("S", plugin, nlp, opts || {});
  const res = solver.call({ x0: M.DM([0, 0]) });
  assertArrayAlmostEqual(res.x.nonzeros(), [1, 2], 3, plugin + ": x");
  assertAlmost(Number(res.f.nonzeros()[0]), 0, 4, plugin + ": f");
}
function test_nlpsol_sqpmethod(M) {
  // SQP needs a QP sub-solver; route through qrqp.
  _nlpsol(M, "sqpmethod", {
    qpsol: "qrqp",
    qpsol_options: { print_iter: false, print_header: false },
    print_iteration: false, print_header: false, print_status: false, print_time: false
  });
}
function test_nlpsol_qrsqp(M) {
  _nlpsol(M, "qrsqp", { print_iteration: false, print_header: false, print_time: false });
}
function test_nlpsol_feasiblesqpmethod(M) {
  _nlpsol(M, "feasiblesqpmethod", {
    qpsol: "qrqp",
    qpsol_options: { print_iter: false, print_header: false },
    print_iteration: false, print_header: false, print_status: false, print_time: false
  });
}
function test_nlpsol_ipopt(M) {
  _nlpsol(M, "ipopt", { ipopt: { print_level: 0, sb: "yes" }, print_time: false });
}
function test_nlpsol_scpgen(M) {
  // scpgen needs MXFunction (calls Function::generate_lifted internally,
  // which is MXFunction-only).  Build the NLP from MX symbolics so the
  // emitted Function is an MXFunction.  Same Rosenbrock-style bowl
  // f = (x-1)^2 + (y-2)^2 as _nlpsol, optimum at (1,2).
  const xy = M.MX.sym("xy", 2);
  const x  = xy.get(false, M.Slice(0n, false));
  const y  = xy.get(false, M.Slice(1n, false));
  const f  = M.plus(M.times(M.minus(x, M.MX(1)), M.minus(x, M.MX(1))),
                    M.times(M.minus(y, M.MX(2)), M.minus(y, M.MX(2))));
  const nlp = { x: xy, f: f };
  const solver = M.nlpsol("S", "scpgen", nlp, {
    qpsol: "qrqp",
    qpsol_options: { print_iter: false, print_header: false },
    print_header: false, print_time: false
  });
  const res = solver.call({ x0: M.DM([0, 0]) });
  assertArrayAlmostEqual(res.x.nonzeros(), [1, 2], 3, "scpgen: x");
  assertAlmost(Number(res.f.nonzeros()[0]), 0, 4, "scpgen: f");
}

runTests([
  ["test_linsol_qr",                test_linsol_qr],
  ["test_linsol_ldl",               test_linsol_ldl],
  ["test_linsol_lsqr",              test_linsol_lsqr],
  ["test_linsol_symbolicqr",        test_linsol_symbolicqr],
  ["test_linsol_tridiag",           test_linsol_tridiag],
  ["test_integrator_rk",            test_integrator_rk],
  ["test_integrator_collocation",   test_integrator_collocation],
  ["test_interpolant_linear",       test_interpolant_linear],
  ["test_interpolant_bspline",      test_interpolant_bspline],
  ["test_rootfinder_newton",        test_rootfinder_newton],
  ["test_rootfinder_fast_newton",   test_rootfinder_fast_newton],
  ["test_rootfinder_implicit_to_nlp", test_rootfinder_implicit_to_nlp],
  ["test_conic_qrqp",               test_conic_qrqp],
  ["test_conic_ipqp",               test_conic_ipqp],
  ["test_conic_nlpsol",             test_conic_nlpsol],
  ["test_conic_highs",              test_conic_highs],
  ["test_nlpsol_sqpmethod",         test_nlpsol_sqpmethod],
  ["test_nlpsol_qrsqp",             test_nlpsol_qrsqp],
  ["test_nlpsol_feasiblesqpmethod", test_nlpsol_feasiblesqpmethod],
  ["test_nlpsol_ipopt",             test_nlpsol_ipopt],
  ["test_nlpsol_scpgen",            test_nlpsol_scpgen],
]);

File: test/julia/plugins.jl
# plugins.jl -- capability port of test/javascript/plugins.js (+ deeper).
# Plugin enumeration / availability / loading / options-introspection across
# every plugin family, plus a trivial construct-and-solve smoke per plugin so
# "it registers" is backed by "it actually computes".  Where the JS twin only
# probes a fixed plugin list, this also exercises the meta enumeration API and
# the doc/options introspection that JS lacks.

const _PL_OD = Dict{String,Any}()

# ---- meta enumeration: plugins() / feature_list() registry strings ----
@testset "CasadiMeta plugins/features enumeration" begin
    pl = C.plugins(C.CasadiMeta)
    @test pl isa AbstractString
    @test occursin("Nlpsol::ipopt", pl)        # registered nlpsol plugin
    @test occursin("Conic::qrqp", pl)           # registered conic plugin
    @test occursin("Integrator::rk", pl)        # registered integrator plugin
    @test occursin("Rootfinder::newton", pl)    # registered rootfinder plugin
    @test occursin("Linsol::qr", pl)            # registered linsol plugin
    fl = C.feature_list(C.CasadiMeta)
    @test fl isa AbstractString
    @test occursin("interface", fl)             # human-readable feature blurbs
end

# ---- availability predicates: assert specific known-present plugins ----
@testset "has_* availability predicates" begin
    @test C.has_nlpsol("ipopt")
    @test C.has_nlpsol("sqpmethod")
    @test C.has_conic("qrqp")
    @test C.has_conic("ipqp")
    @test C.has_integrator("rk")
    @test C.has_integrator("collocation")
    @test C.has_rootfinder("newton")
    @test C.has_rootfinder("kinsol")
    @test C.has_linsol("qr")
    @test C.has_linsol("ldl")
    @test C.has_interpolant("linear")
    @test C.has_interpolant("bspline")
    # negative case: a bogus plugin name reports absent (no throw).
    @test C.has_nlpsol("nope_nope") == false
    @test C.has_conic("nope_nope") == false
    @test C.has_integrator("nope_nope") == false
    @test C.has_rootfinder("nope_nope") == false
    @test C.has_linsol("nope_nope") == false
    @test C.has_interpolant("nope_nope") == false
end

# ---- explicit plugin loading: load_* succeeds for present, throws for bogus ----
@testset "load_* plugin loading" begin
    # load is idempotent for already-registered plugins (warns "already in use").
    @test (C.load_nlpsol("ipopt"); true)
    @test (C.load_conic("qrqp"); true)
    @test (C.load_linsol("qr"); true)
    # bogus plugin: load throws cleanly (dlopen failure surfaced as an exception).
    @test_throws Exception C.load_nlpsol("nope_nope")
    @test_throws Exception C.load_conic("nope_nope")
end

# ---- options / doc introspection (JS twin has none of this) ----
@testset "options & doc introspection" begin
    # nlpsol_options(plugin) lists the option names the solver accepts.
    no = C.nlpsol_options("ipopt")
    @test no isa AbstractVector
    @test "ipopt" in no                        # ipopt's nested-options key
    # conic_options(plugin): qrqp exposes its iteration knobs.
    co = C.conic_options("qrqp")
    @test "max_iter" in co
    @test "print_iter" in co
    # nlpsol_in / nlpsol_out: the standard NLP IO scheme names.
    @test C.nlpsol_in() == ["x0", "p", "lbx", "ubx", "lbg", "ubg", "lam_x0", "lam_g0"]
    @test C.nlpsol_out() == ["x", "f", "g", "lam_x", "lam_g", "lam_p"]
    # doc_nlpsol(plugin): a non-empty documentation blob.
    doc = C.doc_nlpsol("ipopt")
    @test doc isa AbstractString && length(doc) > 100
end

# ---- bogus plugin construction throws cleanly (no segfault) ----
@testset "bogus plugin construction throws" begin
    H = DM([1.0 0.0; 0.0 1.0])
    @test_throws Exception C.conic("x", "nope_nope", Dict("h" => C.sparsity(H)), _PL_OD)
    x = sym(SX, "x")
    @test_throws Exception C.nlpsol("x", "nope_nope", Dict{String,Any}("x" => x, "f" => x * x), _PL_OD)
end

# ---- linsol plugins: A=diag([1,2,3]), x = b ./ [1,2,3] (mirrors plugins.js) ----
function _pl_linsol(plugin)
    A = DM([1.0 0.0 0.0; 0.0 2.0 0.0; 0.0 0.0 3.0])
    ls = C.Linsol("ls", plugin, C.sparsity(A), _PL_OD)
    C.sfact(ls, A); C.nfact(ls, A)
    x = C.solve(ls, A, DM([6.0, 8.0, 9.0]), false)
    @test maximum(abs.(Vector{Float64}(x) .- [6.0, 4.0, 3.0])) < 1e-8
end
for plugin in ["qr", "ldl", "lsqr", "symbolicqr", "csparse", "lapacklu", "lapackqr"]
    (try C.has_linsol(plugin) catch; false end) || continue
    @testset "plugins linsol $plugin" begin _pl_linsol(plugin) end
end

# tridiag needs an actual tri-banded sparsity (diag-only reads OOB) -- mirror JS.
@testset "plugins linsol tridiag" begin
    if !(try C.has_linsol("tridiag") catch; false end)
        @test_skip "tridiag linsol not loaded"
    else
        sp = C.triplet(C.Sparsity, 3, 3, [0, 1, 0, 1, 2, 1, 2], [0, 0, 1, 1, 1, 2, 2])
        A = DM(sp, DM([2.0, 1.0, 1.0, 2.0, 1.0, 1.0, 2.0]))   # tridiag [[2,1,0],[1,2,1],[0,1,2]]
        ls = C.Linsol("ls", "tridiag", sp, _PL_OD)
        C.sfact(ls, A); C.nfact(ls, A)
        x = C.solve(ls, A, DM([4.0, 8.0, 8.0]), false)         # A*[1,2,3] = [4,8,8]
        @test maximum(abs.(Vector{Float64}(x) .- [1.0, 2.0, 3.0])) < 1e-6
    end
end

# ---- integrator plugins: dx/dt = x, x(0)=1 -> x(1) = e ----
function _pl_integrator(plugin, places)
    x = sym(SX, "x")
    F = integrator("F", plugin, Dict("x" => x, "ode" => x), 0.0, 1.0)
    r = F(Dict("x0" => DM(1.0)))
    @test abs(Float64(r["xf"]) - exp(1.0)) < 10.0^(-places)
end
@testset "plugins integrator rk"          begin _pl_integrator("rk", 3) end
@testset "plugins integrator collocation" begin _pl_integrator("collocation", 4) end
for plugin in ["cvodes", "idas"]
    (try C.has_integrator(plugin) catch; false end) || continue
    @testset "plugins integrator $plugin" begin _pl_integrator(plugin, 4) end
end

# ---- interpolant plugins: linear & bspline of f(0)=0,f(1)=1,f(2)=4,f(3)=9 ----
@testset "plugins interpolant linear" begin
    f = C.interpolant("f", "linear", [[0.0, 1.0, 2.0, 3.0]], [0.0, 1.0, 4.0, 9.0], _PL_OD)
    @test abs(Float64(f(DM(1.5))) - 2.5) < 1e-9      # mid of [1,4]
    @test abs(Float64(f(DM(2.0))) - 4.0) < 1e-9      # exact at knot
end
@testset "plugins interpolant bspline" begin
    f = C.interpolant("f", "bspline", [[0.0, 1.0, 2.0, 3.0]], [0.0, 1.0, 4.0, 9.0], _PL_OD)
    @test abs(Float64(f(DM(0.0))) - 0.0) < 1e-6      # clamped left knot
    @test abs(Float64(f(DM(3.0))) - 9.0) < 1e-6      # clamped right knot
end

# ---- rootfinder plugins: x^2 - 2 = 0 from x0=1.5 -> sqrt(2) ----
function _pl_rootfinder(plugin, opts)
    x = sym(SX, "x")
    f = C.Function("f", [x], [C.times(x, x) - 2.0])
    F = C.rootfinder("F", plugin, f, opts)
    @test abs(Float64(F(DM(1.5))) - sqrt(2)) < 1e-6
end
@testset "plugins rootfinder newton"      begin _pl_rootfinder("newton", Dict{String,Any}("print_iteration" => false)) end
@testset "plugins rootfinder fast_newton" begin _pl_rootfinder("fast_newton", _PL_OD) end
@testset "plugins rootfinder kinsol" begin
    (try C.has_rootfinder("kinsol") catch; false end) ?
        _pl_rootfinder("kinsol", Dict{String,Any}("print_level" => 0)) :
        @test_skip "kinsol not loaded"
end
HAS_IPOPT && @testset "plugins rootfinder nlpsol(implicit_to_nlp)" begin
    _pl_rootfinder("nlpsol", Dict{String,Any}("nlpsol" => "ipopt",
        "nlpsol_options" => Dict{String,Any}("print_time" => false,
            "ipopt" => Dict{String,Any}("print_level" => 0, "sb" => "yes"))))
end

# ---- conic plugins: min 0.5(x^2+y^2) - 0.7x - 2.3y -> [0.7, 2.3] ----
function _pl_conic(plugin, opts)
    H = DM([1.0 0.0; 0.0 1.0]); G = DM([-0.7, -2.3])
    S = C.conic("solver", plugin, Dict("h" => C.sparsity(H)), opts)
    r = S(Dict("h" => H, "g" => G, "lbx" => DM([-Inf, -Inf]), "ubx" => DM([Inf, Inf])))
    @test maximum(abs.(Vector{Float64}(r["x"]) .- [0.7, 2.3])) < 1e-4
end
@testset "plugins conic qrqp"  begin _pl_conic("qrqp", Dict{String,Any}("print_iter" => false, "print_header" => false)) end
@testset "plugins conic ipqp"  begin _pl_conic("ipqp", Dict{String,Any}("print_iter" => false, "print_header" => false)) end
@testset "plugins conic highs" begin
    (try C.has_conic("highs") catch; false end) ?
        _pl_conic("highs", Dict{String,Any}("highs" => Dict{String,Any}("output_flag" => false))) :
        @test_skip "highs not loaded"
end
@testset "plugins conic daqp" begin
    (try C.has_conic("daqp") catch; false end) ? _pl_conic("daqp", _PL_OD) : @test_skip "daqp not loaded"
end
@testset "plugins conic qpoases" begin
    (try C.has_conic("qpoases") catch; false end) ?
        _pl_conic("qpoases", Dict{String,Any}("printLevel" => "none")) : @test_skip "qpoases not loaded"
end
HAS_IPOPT && @testset "plugins conic nlpsol(qp_to_nlp)" begin
    _pl_conic("nlpsol", Dict{String,Any}("nlpsol" => "ipopt",
        "nlpsol_options" => Dict{String,Any}("print_time" => false,
            "ipopt" => Dict{String,Any}("print_level" => 0))))
end

# ---- nlpsol plugins: f(x,y) = (x-1)^2 + (y-2)^2 -> (1, 2), f* = 0 ----
function _pl_nlpsol_sx(plugin, opts)
    xy = sym(SX, "xy", 2)
    f = (xy[1] - 1.0)^2 + (xy[2] - 2.0)^2
    S = C.nlpsol("S", plugin, Dict{String,Any}("x" => xy, "f" => f), opts)
    r = S(Dict("x0" => DM([0.0, 0.0])))
    @test maximum(abs.(Vector{Float64}(r["x"]) .- [1.0, 2.0])) < 1e-3
    @test abs(Float64(r["f"])) < 1e-4
end
@testset "plugins nlpsol sqpmethod" begin
    _pl_nlpsol_sx("sqpmethod", Dict{String,Any}("qpsol" => "qrqp",
        "qpsol_options" => Dict{String,Any}("print_iter" => false, "print_header" => false),
        "print_iteration" => false, "print_header" => false, "print_status" => false, "print_time" => false))
end
@testset "plugins nlpsol qrsqp" begin
    (try C.has_nlpsol("qrsqp") catch; false end) ?
        _pl_nlpsol_sx("qrsqp", Dict{String,Any}("print_iteration" => false, "print_header" => false, "print_time" => false)) :
        @test_skip "qrsqp not loaded"
end
@testset "plugins nlpsol feasiblesqpmethod" begin
    (try C.has_nlpsol("feasiblesqpmethod") catch; false end) ?
        _pl_nlpsol_sx("feasiblesqpmethod", Dict{String,Any}("qpsol" => "qrqp",
            "qpsol_options" => Dict{String,Any}("print_iter" => false, "print_header" => false),
            "print_iteration" => false, "print_header" => false, "print_status" => false, "print_time" => false)) :
        @test_skip "feasiblesqpmethod not loaded"
end
HAS_IPOPT && @testset "plugins nlpsol ipopt" begin
    _pl_nlpsol_sx("ipopt", Dict{String,Any}("print_time" => false,
        "ipopt" => Dict{String,Any}("print_level" => 0, "sb" => "yes")))
end

# scpgen needs an MXFunction oracle (generate_lifted is MX-only) -- mirror JS.
@testset "plugins nlpsol scpgen" begin
    if !(try C.has_nlpsol("scpgen") catch; false end)
        @test_skip "scpgen not loaded"
    else
        xy = sym(MX, "xy", 2)
        f = (xy[1] - 1.0)^2 + (xy[2] - 2.0)^2
        S = C.nlpsol("S", "scpgen", Dict{String,Any}("x" => xy, "f" => f),
            Dict{String,Any}("qpsol" => "qrqp",
                "qpsol_options" => Dict{String,Any}("print_iter" => false, "print_header" => false),
                "print_header" => false, "print_time" => false))
        r = S(Dict("x0" => DM([0.0, 0.0])))
        @test maximum(abs.(Vector{Float64}(r["x"]) .- [1.0, 2.0])) < 1e-3
        @test abs(Float64(r["f"])) < 1e-4
    end
end
